build/org.eclipse.pde.build/src/org/eclipse/pde/internal/build/site/PDEState.java

```diff
@@ -509,11 +509,10 @@ public static String querySystemPackages(IExecutionEnvironment environment, Prop
 		}
 		Collection<String> systemPackages;
 		if (releaseVersion <= 8) {
-			IVMInstall eeVM = bestVmInstallFor(environment, vms -> null); // Use selected VM or a perfect match
-			if (eeVM != null) {
-				// In case a VM is selected for an EE, query that VM and use its system-packages
-				systemPackages = querySystemPackages(eeVM, null);
-			} else {
+			systemPackages = vmInstallsFor(environment).filter(environment::isStrictlyCompatible) // Use only selected VM or perfect matches
+					.map(vm -> querySystemPackages(vm, null)) // In case a VM is selected for an EE, query that VM and use its system-packages
+					.filter(Objects::nonNull).findFirst().orElse(null);
```

Comment:
you might want to use
```suggestion
					.filter(Objects::nonNull).findFirst().orElseGet(()->alternative code path);
```

instead of extra check for `null` below.

Reply:
Done.